Developer tools image

glarizza opened the pull request:
This PR covers the work being done to create a unified developer-tools Docker image to be used for local development and in CI pipelines. This image will contain all the tools and dependencies from the existing kitchen-terraform and lint images and ideally will be the only image we need going forward.

glarizza commented:
NOTE: The two initial commits (4b76923 and 78dd70e) will be ported into the developer-tools image in a later commit.

glarizza commented:
* [ ] Move the behavior of `setup_environment()` from ci_integration.sh to `/usr/local/bin/setup_environment.sh`
    * Trap Handler
    * Credentials/SA environment variables for gcloud, gsutil, inspec, and terraform
    * Container should NOT automatically source this file - that will create early trouble like ENTRYPOINT
    * The user (or CI) will source this file first
* [ ] Finish developer-tools image

jeffmccune commented on infra/concourse/build/developer-tools/build/scripts/setup_environment.sh:
```diff
@@ -0,0 +1,64 @@
+  setup_environment
+}
+
+main "$@"
```
This script is conflating up two separate modes of operation from `ci_integration.sh`.  This causes a problem because `set -eu` is enabled in the interactive shell.

The two modes are:

 1. The script is sourced from an interactive shell, e.g. `source setup_environment.sh`.
 2. The script is executed directly, e.g. `./setup_environment.sh`.

Mode 1 is intended for a "library of functions", e.g. "load the library, then setup the environment via `setup_environment()`.

Mode 2 is intended to execute a task, e.g. `"Execute the test lifecycle` here: https://github.com/terraform-google-modules/terraform-google-project-factory/blob/f65c1402fb84932b7a933f50883fdc97dff8d2ee/test/ci_integration.sh#L65

Mode 2 should have the trap handler because the main task should clean up after itself..  Mode 1 should not have a trap handler because when used as a library, the parent shell is responsible for cleanup.

Suggested changes:
 * [ ] Add `if [[ "${BASH_SOURCE[0]}" == "${0}" ]]; then ...` as per https://github.com/terraform-google-modules/terraform-google-project-factory/blob/f65c1402fb84932b7a933f50883fdc97dff8d2ee/test/ci_integration.sh#L72
 * [ ] Implement main()'s behavior as simply executing some command with the correct environment.  So main() sets up the trap handler, sets up the environment, then executes "$@"
 * [ ] Make sure `set -e and set -u` are only enabled in mode 2.

morgante replied:
IMO we should go for the even simpler approach of keeping this script as a "library of functions" and either:
- Always require that one of the functions be explicitly invoked
- Or, have a separate script which is meant to be directly invoked (and can source this script)

glarizza replied:
I've updated the script with the conditional logic, but we can split it out if that's preferred

glarizza replied:
Resolved with the most recent commit

jeffmccune commented on infra/concourse/build/developer-tools/build/scripts/setup_environment.sh:
```diff
@@ -0,0 +1,64 @@
+fi
+
+# Always clean up.
+DELETE_AT_EXIT="$(mktemp -d)"
```
I know this comes from `ci_integration.sh`, but it should move into main() so the tempdir isn't created when used as a library.

glarizza replied:
Done!

jeffmccune commented on infra/concourse/build/developer-tools/build/scripts/setup_environment.sh:
```diff
@@ -0,0 +1,64 @@
+DELETE_AT_EXIT="$(mktemp -d)"
+
+finish() {
+  [[ -d "${DELETE_AT_EXIT}" ]] && rm -rf "${DELETE_AT_EXIT}"
```
This should be `${DELETE_AT_EXIT:-}` in case the variable hasn't been set from `main()`.

glarizza replied:
Is this still valid with the scripts being split and the variable always being set to `$(mktemp -d)`?

jeffmccune commented on infra/concourse/build/developer-tools/build/scripts/setup_environment.sh:
```diff
@@ -0,0 +1,64 @@
+  [[ -d "${DELETE_AT_EXIT}" ]] && rm -rf "${DELETE_AT_EXIT}"
+}
+
+setup_environment() {
```
Looks good!

jeffmccune commented on infra/concourse/build/developer-tools/build/scripts/setup_environment.sh:
```diff
@@ -0,0 +1,64 @@
+  trap finish EXIT
+
+  # Setup environment
+  setup_environment
```
After setting up the environment let's have main call `"$@"` so an arbitrary command may be run with the credentials initialized.  For example, make calls `setup_environment.sh kitchen create` which will initialize credentials, call kitchen create, then clean up.

jeffmccune commented:
Agreed, that will clarify the intent of each.

glarizza commented:
Commit e6ec294 splits `setup_enviroment.sh` into two separate scripts to address each operation mode. `execute_with_credentials.sh` addresses operation mode 2 and `setup_environment.sh` addresses operation mode 1.

glarizza commented:
Note that in https://github.com/GoogleCloudPlatform/cloud-foundation-toolkit/pull/250/files#diff-8d485d6917e30199e54c9f74cea3e2f9R9 the version of this image has been set to `0.0.1` for testing - let me know what version we should use for release and I'll amend that.

morgante commented on infra/concourse/build/developer-tools/build/scripts/execute_with_credentials.sh:
```diff
@@ -0,0 +1,37 @@
+DELETE_AT_EXIT="$(mktemp -d)"
+export TMPDIR="${DELETE_AT_EXIT}"
+
+if [[ -z "${SERVICE_ACCOUNT_JSON}" ]]; then
```
This looks duplicative. Can't this file just source setuo_environment and exec a function from there?

glarizza replied:
The environment variable check is duplicated, and I did that to set expectations for anyone using this script (and to fail early before we got to `setup_environment.sh`). I don't have a problem removing it and deferring to the original

glarizza replied:
Addressed in a5a6fce

jeffmccune commented on infra/concourse/build/developer-tools/Dockerfile:
```diff
@@ -0,0 +1,78 @@
+    kubectl version --client=true
+
+# Cleanup intermediate build artifacts
+RUN rm -rf /build
```
Suggest overriding the entrypoint or CMD or whatever is causing this behavior:

```
docker run --rm -it cft/developer-tools:0.0.1
irb(main):001:0>
```

I prefer dropping to an interactive bash shell since this is how many of us work with the developer tools image based on the quick survey we took in our video call.

glarizza replied:
Addressed in 5e2901e

jeffmccune commented on infra/concourse/build/developer-tools/Dockerfile:
```diff
@@ -0,0 +1,80 @@
+
+# Cleanup intermediate build artifacts
+RUN rm -rf /build
+ENTRYPOINT ["/bin/bash"]
```
I'm not sure this is working, now I get this:

```
docker run --rm -e SERVICE_ACCOUNT_JSON -it -v `pwd`:/workspace cft/developer-tools:0.0.1 bash
/bin/bash: /bin/bash: cannot execute binary file
```

glarizza replied:
Yeah, that's definitely not the solution. I'll revert it locally and see if I can figure out where IRB is coming in

glarizza replied:
Corrected with f58bbcfa9855734f7143691affc8e3650bc69792

jeffmccune commented on infra/concourse/build/developer-tools/build/scripts/task_wrapper_scripts/check_bash:
```diff
@@ -0,0 +1,17 @@
+# limitations under the License.
+
+
+source /usr/local/bin/task_helper_functions.sh && check_bash
```
We need to look into this, I get what looks to be an infinite recursion bug:

```bash
docker run --rm -e SERVICE_ACCOUNT_JSON -it -v `pwd`:/workspace cft/developer-tools:0.0.1
bash-5.0# source /usr/local/bin/task_helper_functions.sh
bash-5.0# cd /workspace/
bash-5.0# check_bash
/bin/bash: warning: shell level (1000) too high, resetting to 1
/bin/bash: warning: shell level (1000) too high, resetting to 1
/bin/bash: warning: shell level (1000) too high, resetting to 1
/bin/bash: warning: shell level (1000) too high, resetting to 1
^C
bash-5.0#
```

jeffmccune replied:
Fixed it in 39c8b0ee8b9a420638c01dc34d103932abdf05fc Remove bugged check_bash, use check_shell instead

jeffmccune commented on infra/concourse/build/developer-tools/build/scripts/task_helper_functions.sh:
```diff
@@ -0,0 +1,238 @@
+  echo "Checking file headers"
+  # Use the exclusion behavior of find_files
+  find_files . -type f -print0 \
+    | compat_xargs -0 python test/verify_boilerplate.py
```
Need to fix:

```
bash-5.0# check_headers
Checking file headers
python: can't open file 'test/verify_boilerplate.py': [Errno 2] No such file or directory
```

jeffmccune replied:
Fixed

jeffmccune commented:
@glarizza I squashed all of the commits into one and saved the branch from the end of our day to 4bc34a1 -> devex2.0/developer_tools_2019-07-24_1822

@morgante @aaron-lane This is ready for review / merge.  Would be good to get it soon to keep the patch size down if possible.

jeffmccune commented:
Ready to go, I'm no longer pushing until reviewed.

morgante commented on infra/concourse/build/developer-tools/build/install_dependencies.sh:
```diff
@@ -0,0 +1,69 @@
+# unclear why perl is needed, but is good to have
+apk add --no-cache perl
+
+# python 2 is needed for compatibility and linting
```
We should be moving to Python 3 for everything.

glarizza replied:
Understood. No action taken per chat

morgante commented on infra/concourse/build/developer-tools/Makefile:
```diff
@@ -0,0 +1,87 @@
+		cft/developer-tools:0.0.1 \
+		make test_lint
+
+.PHONY: test_lint
```
Instead of copying "test_lint" into each repo we should provide a single command which can be invoked to do everything.

jeffmccune replied:
How about this as the interface between a module and the developer-tools container image?

```
.PHONY: docker_test_lint
docker_test_lint:
	docker run --rm -it \
		-v $(CURDIR):/workspace \
		cft/developer-tools:0.0.1 \
		test_lint
```

If inside the container, the single command would be `test_lint`

morgante replied:
Yes that make sense.

glarizza replied:
Updated

morgante commented:
Issues noticed:
- [ ] Expects us to copy too much code into all Makefiles: [details](https://github.com/GoogleCloudPlatform/cloud-foundation-toolkit/pull/250#discussion_r308931925)
- [ ] It looks like `make test_lint` exits immediately after the first failed linter. It should run all linters every time

glarizza commented:
* [ ] Ensure `terraform_docs.sh` shim is being used in the image

glarizza commented:
> Issues noticed:
> 
> * [x]  Expects us to copy too much code into all Makefiles: [details](https://github.com/GoogleCloudPlatform/cloud-foundation-toolkit/pull/250#discussion_r308931925)
> * [ x]  It looks like `make test_lint` exits immediately after the first failed linter. It should run all linters every time
> * [ x]  It looks like linters are traversing into `.terraform` directories. They never should.
> * [ ]  The whitespace linter is failing without explaining why.

I can't edit or click the checkboxes in Morgante's comment so I'm quoting it here to track my progress

morgante commented:
Looks like I found another issue where the linters are still exiting early. Specifically the missing newline check.

```
docker run --rm -it \
		-v /Users/morgantep/code/google/foundation/modules/kms:/workspace \
		gcr.io/cloud-foundation-cicd/cft/developer-tools:0.0.1 \
		/usr/local/bin/test_lint.sh
Checking for trailing whitespace
Checking for missing newline at end of file
Error: No newline at end of file ./Makefile
make: *** [docker_test_lint] Error 1
```

glarizza commented:
Fixed the early-exit issue - it was because of `set -eu` in a module-specific `task_helper_functions.sh` file. That file was sourced AFTER `set +e` in `test_lint.sh` so the flag was overwritten. I changed the order of those two operations so `set +e` will always override anything from the module-specific `task_helper_functions.sh` file for `test_lint.sh`

morgante commented:
Another thing we should add/restore: the linters should check if `generate_docs` has been run.

glarizza commented:
Updated the PR last night with the `check_documentation` function that gets run from `test_lint.sh`. This function will determine whether `generate_docs` needs to be run on the module code at `/workspace` and will raise an error in response.

glarizza commented:
Unfortunately I have to transition to other work, but the checklists I had access to (and the subsequent comments) are all up to date. The 0.0.1 image has been released as of my last commit, and has been used for testing with the Network module PR (https://github.com/terraform-google-modules/terraform-google-network/pull/58) and the KMS PR that implemented Cloud Build (https://github.com/terraform-google-modules/terraform-google-kms/pull/11/files).
